detector/zoo/detect.go

```diff
@@ -187,57 +191,97 @@ func (md *MasterDetector) notifyAllMasters(path string, list []string, obs detec
 // once, and each time the spec'd listener will be added to the list of those receiving notifications.
 func (md *MasterDetector) Detect(f detector.MasterChanged) (err error) {
 	// kickstart zk client connectivity
-	md.bootstrap.Do(func() { go md.client.connect() })
+	if err := func() (e error) {
+		log.V(2).Infoln("invoking detector boostrap")
+		md.bootstrapLock.Lock()
+		defer md.bootstrapLock.Unlock()
+
+		clientConfigured := md.client != nil
+		if e = md.bootstrapFunc(); e == nil && !clientConfigured && md.client != nil {
+			// chain the lifetime of this detector to that of the newly created client impl
+			client := md.client
+			md.cancel = client.stop
+			go func() {
+				defer close(md.done)
+				<-client.stopped()
+			}()
+		}
+		return
+	}(); err != nil {
+		log.V(3).Infoln("failed to execute bootstrap function", err.Error())
+		return err
+	}
 
 	if f == nil {
 		// only ever install, at most, one ignoreChanged listener. multiple instances of it
 		// just consume resources and generate misleading log messages.
 		if !atomic.CompareAndSwapInt32(&md.ignoreInstalled, 0, 1) {
+			log.V(3).Infoln("ignoreChanged listener already installed")
 			return
 		}
 		f = ignoreChanged
 	}
 
+	log.V(3).Infoln("spawning detect()")
 	go md.detect(f)
 	return nil
 }
 
 func (md *MasterDetector) detect(f detector.MasterChanged) {
+	log.V(3).Infoln("detecting children at", currentPath)
 detectLoop:
 	for {
-		started := time.Now()
 		select {
 		case <-md.Done():
 			return
-		case <-md.client.connections():
-			// we let the golang runtime manage our listener list for us, in form of goroutines that
-			// callback to the master change notification listen func's
-			if watchEnded, err := md.client.watchChildren(currentPath, ChildWatcher(func(zkc *Client, path string) {
-				md.childrenChanged(zkc, path, f)
-			})); err == nil {
-				log.V(2).Infoln("detector listener installed")
+		default:
+		}
+		log.V(3).Infoln("watching children at", currentPath)
+		path, childrenCh, errCh := md.client.watchChildren(currentPath)
+		rewatch := false
+		for {
+			started := time.Now()
+			select {
+			case children := <-childrenCh:
+				md.childrenChanged(path, children, f)
+			case err, ok := <-errCh:
+				// check for a tie first (required for predictability (tests)); the downside of
+				// doing this is that a listener might get two callbacks back-to-back ("new leader",
+				// followed by "no leader").
 				select {
-				case <-watchEnded:
+				case children := <-childrenCh:
+					md.childrenChanged(path, children, f)
+				default:
+				}
+				if ok {
+					log.V(1).Infoln("child watch ended with error, signaling master lost; error was:", err.Error())
```

Comment:
remove the word signalling, because we don't always do it (for example, if the leader is already == "")